### ISSUE TITLE
[Test] not working with ExpectedResult and [ValueSource]

### ISSUE THREAD

**Issue by mknejp** [nunit/nunit#1194]
With the NuGet 3.0.1 version, given the following tests:
```cs
[Test(ExpectedResult = true)]
public bool Foo1([ValueSource(nameof(FooValues))] bool result)
{
    return result;
}

[Test(ExpectedResult = true)]
public bool Foo2([ValueSource(nameof(FooValues))] bool dummy)
{
    return true;
}

static bool[] FooValues = new bool[] { true };
```
I get this output from `nunit3-console`
```txt
NUnit Console Runner 3.0.5813
Copyright (C) 2015 Charlie Poole

Runtime Environment
   OS Version: Microsoft Windows NT 6.3.9600.0
  CLR Version: 4.0.30319.42000

Test Files
    x.dll


Errors and Failures

1) Invalid : x.Foo1(True)
Method has non-void return value, but no result is expected

2) Invalid : x.Foo2(True)
Method has non-void return value, but no result is expected

Run Settings
    WorkDirectory: x
    NumberOfTestWorkers: 8

Test Run Summary
    Overall result: Failed
   Tests run: 10, Passed: 10, Errors: 0, Failures: 0, Inconclusive: 0
     Not run: 2, Invalid: 2, Ignored: 0, Explicit: 0, Skipped: 0
  Start time: 2016-01-14 21:58:25Z
    End time: 2016-01-14 21:58:25Z
    Duration: 0,103 seconds

Results (nunit3) saved as TestResult.xml
```
Is this a bug or by design? Removing the parameters (and therefore `ValueSource`) works fine. It would be really convenient to be able to feed variables from a source and simply test that they all return the same result. The actual test I wanted to do this with has multiple parameters and checks that a certain operation returns the same result for all combinations. Sure, I could put `Assert.That(..., Is.EqualTo(...))` in the function, which I'll have to do for now, but using `ExpectedResult` seems more convenient (and less likely to screw up the assert statement).

**Comment by CharliePoole** [comment]
I think this is both a bug and by design. :-)

ExpectedResult was originally added for parameterized tests. For each test case, you could specify the result that should be returned instead of providing it as an argument, which  would have to be tested by an Assert in the test code. As you point out, it's convenient.

Later, in response to a user request, we extended ExpectedResult for use on non-parameterized tests, which is why you can specify it on a TestAttribute. This too is convenient, but not very common.

In the case of ValueSource, however, the user does not specify the test cases. Instead, NUnit creates them by making various combinations of the provided arguments.  There is no way for NUnit to determine what should be the expected result of a particular combination. The attribute is therefore ignored in this case. It would also be ignored on any other type of parameterized test, using Values, TestCase, TestCaseSource, etc.

When using TestCase or TestCaseSource, you have the ability to specify the ExpectedResult. For ValueSource there is no clear way to indicate which combinations of attributes should generate it.

That's the by design part... We could consider this a bug in two ways...

1. We could say that **all** the tests under your TestAttrbute should be expected to return the same thing. This seems both not too useful and somewhat tricky to implement. Consider that the  TestCases might specify their own, conflicting ExpectedResult.

2. We can give an error message when an ExpectedResult is specified on a TestAttribute for any kind of parameterized test. That would be my preference.

What do you think?

**Comment by CharliePoole** [comment]
I should add that the error you are getting is similar to my option 2. It just is not worded in a clear enough way.

**Comment by mknejp** [comment]
I understand the difference between `TestCase(Source)` and `Test`. I would have used `TestCase(Source)` if every run had a different result which is not the case here. But implementing a `TestCaseSource` is pretty verbose already.

I was obviously expecting option 1.

I do not see the conflict with `TestCase` as to my understanding every `TestCase` **must have** its own `ExpectedResult` if the return is non-void (which is the error I get when adding an additional `[TestCase(true)]` to my example).

So the `ExpectedResult` from `Test` does not seem to carry over to `TestCase` instances. Therefore for option 1 the `ExpectedResult` of a parameterized `Test` should only apply to the test cases generated by the affected `TestAttribute`.

As for the "not too useful" part, that is a difficult call to make. In my case I am testing the relationship between different objects, where all provided combinations must have the same relationship, so it seemed natural to me to use `ValueSource` for providing the objects and `ExpectedResult` to ensure the relationship is always the same.

**Comment by CharliePoole** [comment]
I see. Your example methods have only one parameter, but I guess that was just to illustrate the problem. I can see that it would help you to have the TestAttribute's ExpectedResult apply to every method.

On the negative side, while you are very clear on this, many people get pretty confused about attributes that apply to a parameterized test method equally for all cases. Let's see what other comments you get on this.

On the plus side, if you have multiple combinations of data, which should all have the same result, you may be able to implement it more clearly by simply including the expected result as a constant, or as a code expression, right in the test. Here is a simplistic example:

[Test]
AdditionIsCommutative([Values(1,2,3)]int x, [Values(-100,0,100)]int y)
{
    Assert.That(x+y, Is.EqualTo(y+x));
}

If I wanted to weed out bad data with Assume and/or use DataPoint for the data, then I would make it a Theory.

**Comment by mknejp** [comment]
I have updated the example to multiple parameters.

Using Assert.That is what'll have to use for now since my original idea doesn't currently work. I'd prefer the attribute solution as it is less verbose and there is less room for user error. It's also more self-describing and not hidden somewhere in the test method. I guess that can be filed under "personal preference".

**Comment by CharliePoole** [comment]
As you say, it may be a matter of personal preference or style. FWIW here's how I would do your example...

```
[Test]
public bool LessThan([ValueSource(nameof(X))] int x, [ValueSource(nameof(Y))] int y)
{
    Assert.That( x < y );
}

static int[] X = new int[] { 1, 2, 3, 4 };
static int[] Y = new int[] { 5, 6, 7, 8 };
```

Stylistically, what I like is that the test method iscompletely self-contained. There is no need for any external attribute to explain what it does. And it's a few characters shorter. :-)

I didn't know what you meant by "Sequence" as an argument tot he test attribute, so I left it out.

**Comment by CharliePoole** [comment]
I changed the description to reflect what I think you are asking for and labelled this as an "idea." By that, we mean it's something we may or may not do and which requires some discussion to decide.

**Comment by CharliePoole** [comment]
@nunit/contributors Another issue that has been sitting in "Discussion" for too long. Let's decide what to do and move it ahead. The options seem to be:

1. The simple one of giving an error if TestAttribute has an ExpectedResult on a parameterized method.

2. The more complex one of making it work as proposed. Note that this includes dealing with a mixture of data sources, some of which may supply a conflicting ExpectedResult. One possible way around this is to give an error if ExpectedResult is provided on more than one attribute of the method.

My own preference is the first approach.

**Comment by rprouse** [comment]
I vote for the first approach. I can't see many cases where one expected result would work for every combination of the input values. To be honest, the example code looks a bit contrived.

**Comment by AndyPook** [comment]
Associated with this is that no attributes are applied to the child tests. In our case we want to be able to apply `[NonParalleizable]`. It was a surprise to see these tests running concurrently. Seems to be due to `NUnitTestCaseBuilder`:64-65 only applying the methods attributes if there are no parameters. There are a lot of attributes that implement `IApplyToTest`, so none of these are being applied to these tests.

I assume that there is a reason for this?
What would be consequence of always using `ApplyAttributesToTest` (ie deleting line 64)?

**Comment by CharliePoole** [comment]
@AndyPook @jnm2 
I agree it's an associated issue and we should discuss it. Let's not do it here!!!

Reason: This started out as an "idea" for discussion. Rob shepherded it to the point where it proposes a specific solution to a very narrow problem (ExpectedResult property on TestAttribute). It's low priority and so has not been done yet, but it's ready for somebody to work on at any point.

If we start a big discussion of "related" stuff here, we lose the ground we gained in the earlier discussion! 

If we were a physically co-located team, it would be easier to separate these broad discussions from the solving of specific bugs. For example, we might gather around the whiteboard in a conference room to brainstorm various ways to handle attributes on parameterized methods. That's not possible for us. Since we rely almost entirely on GitHub to organize our interactions, it's pretty important that we respect issue boundaries, labels, etc. Let's leave this as an `is:bug` item (i.e. ready to work on) rather than injecting other stuff that makes it non-atomic and forces it to become `is:idea` and therefore tell people not to work on the code.

**Comment by AndyPook** [comment]
Fair enough. I do understand that this is a passion project, not like I'm paying for support!

One question though, looking for a little insight...
The condition at `NUnitTestCaseBuilder`:64, obviously put there for a reason, is there a simple explanation for for needing it? Why don't we want the `IApplyToTest` attrs applied to the generated tests? I'm assuming there's some other scenario where it makes sense?

Thanks for your continued efforts on nunit

**Comment by OsirisTerje** [comment]
This issue seems to have been fixed somewhere between 3.0.1 and latest 4.6.1.

When running `dotnet test` on the code above, we get:

<img width="2000" height="157" alt="Image" src="https://github.com/user-attachments/assets/778bc342-b723-4c76-bc4f-eabcaba77720" />
